src/filesystem/writer.rs

```diff
@@ -22,30 +23,126 @@ use crate::{
-    pub block_size: u32,
-    /// See [`SuperBlock`].`block_log`
-    pub block_log: u16,
-    /// See [`SuperBlock`].`compressor`
-    pub compressor: Compressor,
-    /// See [`Squashfs`].`compression_options`
-    pub compression_options: Option<CompressionOptions>,
-    /// See [`SuperBlock`].`mod_time`
-    pub mod_time: u32,
-    /// See [`Squashfs`].`id`
-    pub id_table: Option<Vec<Id>>,
+    pub(crate) kind: Kind,
+    /// The size of a data block in bytes. Must be a power of two between 4096 (4k) and 1048576 (1 MiB).
+    pub(crate) block_size: u32,
+    /// Last modification time of the archive. Count seconds since 00:00, Jan 1st 1970 UTC (not counting leap seconds).
+    /// This is unsigned, so it expires in the year 2106 (as opposed to 2038).
+    pub(crate) mod_time: u32,
+    /// 32 bit user and group IDs
+    pub(crate) id_table: Vec<Id>,
     /// Information for the `/` node
-    pub root_inode: SquashfsDir,
+    pub(crate) root_inode: SquashfsDir,
+    /// Compressor used when writing
+    pub(crate) compressor: FilesystemCompressor,
     /// All files and directories in filesystem, including root
-    pub nodes: Vec<Node<SquashfsFileWriter<'a, R>>>,
+    pub(crate) nodes: Vec<Node<SquashfsFileWriter<'a, R>>>,
+    /// The log2 of the block size. If the two fields do not agree, the archive is considered corrupted.
+    pub(crate) block_log: u16,
+}
+
+impl Default for FilesystemWriter<'_> {
+    /// Create default FilesystemWriter
+    ///
+    /// block_size: `DEFAULT_BLOCK_SIZE`, compressor: default XZ compression, no nodes,
+    /// kind: `Kind::default()`, and mod_time: `0`.
+    fn default() -> Self {
+        let block_size = DEFAULT_BLOCK_SIZE;
+        Self {
+            block_size,
+            mod_time: 0,
+            id_table: vec![],
+            root_inode: SquashfsDir {
+                header: NodeHeader::default(),
+            },
+            compressor: FilesystemCompressor::default(),
+            kind: Kind::default(),
+            nodes: vec![],
+            block_log: (block_size as f32).log2() as u16,
+        }
+    }
 }
 
 impl<'a, R: ReadSeek> FilesystemWriter<'a, R> {
-    /// use the same configuration then an existing SquashFsFile
+    /// Set block size
+    pub fn set_block_size(&mut self, block_size: u32) {
+        self.block_size = block_size;
+        self.block_log = (block_size as f32).log2() as u16;
+    }
+
+    /// Set time of image as `mod_time`
+    pub fn set_time(&mut self, mod_time: u32) {
+        self.mod_time = mod_time;
+    }
+
+    /// Set time of image as current time
+    pub fn set_current_time(&mut self) {
+        self.mod_time = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs() as u32;
+    }
+
+    /// Set kind as `kind`
+    pub fn set_kind(&mut self, kind: Kind) {
+        self.kind = kind;
+    }
+
+    /// Set root header as `header`
+    pub fn set_root_header(&mut self, header: NodeHeader) {
```

Comment:
Change to `set_mode` since we only use this internally to set the permission bits of the root for.